docs/proposals/04-etcd-member-custom-resource.md

```diff
@@ -0,0 +1,545 @@
+
+[`dbSizeInUse`](https://github.com/etcd-io/etcd/blob/a603c0798948941d453b158af794edab1a8230be/mvcc/backend/backend.go#L65-L68) - logical storage space used by the etcd database, not accounting for free pages in the DB due to etcd history compaction
+
+The leading-backup-sidecar performs periodic defragmentations of the DBs of all the etcd-members in the cluster, controlled via a defragmentation cron schedule provided to each backup-sidecar. Defragmentation is a costly maintenance operation and causes a brief downtime to the etcd-member being defragmented, due to which the leading-backup-sidecar defragments each etcd-member sequentially. This ensures that only one etcd-member would be unavailable at any given time, thus avoiding an accidental quorum loss in the etcd cluster.
+
+The authors propose to move the responsibility of orchestrating these individual defragmentations to druid due to the following reasons:
+
+- Since each backup-sidecar only has knowledge of the health of its own etcd, it can only determine whether its own etcd can be defragmented or not, based on etcd-member health. Trying to defragment a different healthy etcd-member while another etcd-member is unhealthy would lead to a transient quorum loss.
+- Each backup-sidecar is only a `sidecar` to its own etcd-member, and by good design principles, it must not be performing any cluster-wide maintenance operations, and this responsibility should remain with the etcd cluster operator.
+
+Additionally, defragmentation of an etcd DB becomes inevitable if the DB size exceeds the specified DB space quota, since the etcd DB then becomes read-only, ie no write operations on the etcd would be possible unless the etcd DB is defragmented and storage space is freed up. In order to automate this, druid will now need information about the etcd DB size from each member, specifically the leading etcd-member, so that a cluster-wide defragmentation can be triggered if the DB size reaches a certain threshold, as already described by [this issue](https://github.com/gardener/etcd-backup-restore/issues/556).
+
+The authors propose to enhance each etcd-member to regularly publish information about the `dbSize` and ` dbSizeInUse` so that druid may trigger defragmentation for the etcd cluster. 
+
+```yaml
+dbSize: <db-size> # e.g 6Gi
+dbSizeInUse: <db-size-in-use> # e.g 3.5Gi
+```
+
+Difference between `dbSize` and `dbSizeInUse` gives a clear indication of how much storage space would be freed up if a defragmentation is performed. If the difference is not significant (based on a configurable threshold provided to druid), then no defragmentation should be performed. This will ensure that druid does not perform frequent defragmentations that do not yield much benefit. Effectively it is to maximise the benefit of defragmentation since this operations involves transient downtime for each etcd-member. 
```

Comment:
Will the `defragmentationSchedule` still be considered for defragmentation? If not, I'm a bit worried about single-node clusters.
While mutli-node clusters it's possible to execute defragmentation in a rolling fashion and more precautionary, it's a difficult balance for single-node clusters because of the involved downtime.

Reply:
Yes defragmentation will now happen in two ways:

- As per the configured schedule. Today it is set to once in 24 hrs. This will always be in place.
- Threshold based defragmentations - this will be triggered as explained and will ensure that etcd DB size is kept in check even during sudden/long peaks which causes the etcd DB to increase in size rapidly. There is auto-compaction but that is not sufficient to release memory and must be followed by defragmentation.

Reply:
Sounds good!